mlflow/projects/yarn.py

```diff
@@ -0,0 +1,284 @@
+        )
+    )
+
+    if user:
+        spec.user = user
+
+    if queue:
+        spec.queue = queue
+
+    if node_label:
+        service.node_label = node_label
+
+    return skein_client.submit(spec)
+
+
+def _generate_skein_service(memory, num_cores, num_containers, env_vars, additional_files,
+                            hadoop_conf_dir, pex_env, module_name, args):
+    env = dict(env_vars) if env_vars else dict()
+    env.update(
+        {
+            'SKEIN_CONFIG': './.skein',
+            'PEX_ROOT': "./.pex",
+            "PYTHONPATH": ".",
+            "GIT_PYTHON_REFRESH": "quiet"
+        }
+    )
+    launch_options = "-m" if not module_name.endswith(".py") else ""
+    launch_args = args if args else ""
+
+    dict_files_to_upload = {os.path.basename(path): os.path.abspath(path)
+                            for path in additional_files}
+    python_bin = "./%s" % os.path.basename(pex_env) if pex_env.endswith(
+        '.pex') else "./%s/bin/python" % os.path.basename(pex_env)
+
+    return skein.Service(
+        resources=skein.model.Resources(memory, num_cores),
+        instances=num_containers,
+        files=dict_files_to_upload,
+        env=env,
+        script="""
+                    set -x
+                    env
+                    export HADOOP_CONF_DIR=%s
+                    %s %s %s %s
+                """ % (hadoop_conf_dir, python_bin, launch_options, module_name, launch_args)
+    )
+
+
+def _validate_yarn_env(project):
+    if not project.name:
+        raise ExecutionException("Project name in MLProject must be specified when "
+                                 "using Yarn backend.")
+    if not project.pex_env:
+        raise ExecutionException("Project must specify pex environment in order to run "
+                                 "on Yarn backend.")
+
+
+def _merge_env_lists(env_params, env_yarn_cfg):
+    env = dict(value.split('=') for value in env_params)
+    env.update(dict(value.split('=') for value in env_yarn_cfg))
+    return env
+
+
+def _get_key_from_params(params, key, remove_key=True):
+    if key not in params:
+        return []
+
+    values = params[key].split(',')
+    if remove_key:
+        del params[key]
+
+    return values
+
+
+def _parse_yarn_config(backend_config, extra_params=None):
+    """
+    Parses configuration for yarn backend and returns a dictionary
+    with all needed values. In case values are not found in ``backend_config``
+    dict passed, it is filled with the default values.
+    """
+
+    extra_params = extra_params or {}
+    if not backend_config:
+        raise ExecutionException("Backend_config file not found.")
+    yarn_config = backend_config.copy()
+
+    for cfg_key in [YARN_NUM_CORES, YARN_MEMORY, YARN_QUEUE,
+                    YARN_HADOOP_FILESYSTEMS, YARN_HADOOP_CONF_DIR,
+                    YARN_ENV, YARN_ADDITIONAL_FILES]:
+        if cfg_key not in yarn_config.keys():
+            yarn_config[cfg_key] = extra_params.get(YARN_NUM_CORES,
```

Comment:
it should be `extra_params.get(cfg_key, ...)` no?... otherwise we are always looking for YARN_NUM_CORES.

would be nice to catch this in tests.

Reply:
Good catch, I added a test for this